### ISSUE TITLE
Add functions for conversion between Date and Double

### ISSUE THREAD

**Issue by romildo** [qrilka/xlsx#92]
`Date` fields in worksheets are available as `Double`. Please, add functions to convert from `Double` to a `Date` type and vice-versa.

**Comment by qrilka** [comment]
thanks @romildo for reporting this, actually I thought about this almost from the time I started working on this library but probably didn't have much motivation to work on it. If you have something ready - I'll be glad to have a PR otherwise I will start looking into it soon.

**Comment by stla** [comment]
I use this module, requiring the `dates` library.

```
{-# LANGUAGE OverloadedStrings #-}
module ExcelDates
  where
import           Data.Dates
import           Data.Dates.Formats (parseDateFormat)
import           Data.Either.Extra  (fromRight')
import           Data.Text          (Text)
import qualified Data.Text          as T
import           Text.Printf        (printf)

-- | convert a date given as "YYYY/MM/DD" to its corresponding Excel number value
excelDate :: String -> Double
excelDate date = fromIntegral $ datesDifference origin datetime
    where origin = DateTime{year = 1899, month=12, day=30, hour=0, minute=0, second=0}
          datetime = fromRight' $ parseDateFormat "YYYY/MM/DD" date

-- | convert an integer to a date formatted as "YYYY-MM-DD"
intToDate :: Integer -> Text
intToDate n = T.intercalate "-" $ map (\x -> T.pack (printf "%02d" x :: String)) [year date, month date, day date]
  where origin = DateTime{year = 1899, month=12, day=30, hour=0, minute=0, second=0}
        date = addInterval origin (Days n)
```

**Comment by romildo** [comment]
I have written the following to convert a `Double` value to an `UTCTime` value:

``` haskell
module ExcelDate where

import Data.Time.Calendar (Day, fromGregorian, addDays, diffDays)
import Data.Time.Clock (UTCTime (UTCTime), picosecondsToDiffTime, diffTimeToPicoseconds)

origin :: Day
origin = fromGregorian 1899 12 30

dateFromDouble :: RealFrac t => t -> UTCTime
dateFromDouble d = UTCTime day diffTime
  where
    (numberOfDays, fractionOfOneDay) = properFraction d
    day = addDays numberOfDays origin
    diffTime = picosecondsToDiffTime (round (fractionOfOneDay * 24*60*60*1E12))

dateToDouble :: Fractional a => UTCTime -> a
dateToDouble (UTCTime day diffTime) = numberOfDays + fractionOfOneDay
  where
    numberOfDays = fromIntegral (diffDays day origin)
    fractionOfOneDay =  fromIntegral (diffTimeToPicoseconds diffTime) / (24*60*60*1E12)
```

Example of use:
```
$ ghci /tmp/ExcelDate
![spreadsheet](https://user-images.githubusercontent.com/1217934/27974861-279dd678-6336-11e7-89c6-acfa9077bb0b.png)
![spreadsheet](https://user-images.githubusercontent.com/1217934/27974895-5b2f45ee-6336-11e7-8b45-a32f9aea78c2.png)

.hs 
GHCi, version 8.0.2: http://www.haskell.org/ghc/  :? for help
package flags have changed, resetting and loading new packages...
Loaded GHCi configuration from /home/romildo/.ghc/ghci.conf
[1 of 1] Compiling ExcelDate        ( /tmp/test.hs, interpreted )
Ok, modules loaded: ExcelDate.

*ExcelDate> :m + Data.Time

*ExcelDate Data.Time> let Just oneDate = parseTimeM True defaultTimeLocale "%F %T" "2017-07-13 18:00:00" :: Maybe UTCTime

*ExcelDate Data.Time> oneDate 
2017-07-13 18:00:00 UTC

*ExcelDate Data.Time> dateToDouble oneDate
42929.75

*ExcelDate Data.Time> dateFromDouble 42929.75
2017-07-13 18:00:00 UTC
```

@qrilka
As I have just started using `xlsx`, I do not know much of its structure. So I do not know the best place for this code. With some hint I can write a PR.

**Comment by qrilka** [comment]
@romildo I like that code. It look like Excel is a bit more complicated than this but adding 1 more base should be easy (addin "base" parameter but we'll need also to have that workbook property being supported - I could add it on top of your PR), quoting the spec (it's a pity that ECMA publishes only pdfs):

> ### 18.17.4.1 Date Conversion for Serial Date-Times
> A serial date-time is a number that represents a date and time. This signed value is in units of days relative to the base date for the selected date system. Serial date-times increase by 1 into each successive day and decrease by 1 into each preceding day. Fractional portions of serial date-times represent fractions of a single day. [Example: When using the 1900 date system, which has a base date of 30 th December 1899, a serial datetime of 1.5 represents midday on the 31 st December 1899 (serial date-time day 1), or 1899-12-31T12:00. A serial date-time of -4.25 represents 6 pm on the 25 th December 1899, or 1899-12-25T18:00. end example] The base dates and the related serial date-times represent local date and time.
> Two different bases are used for converting dates to and from serial date-times:
> * In the 1900 date system, the lower limit is January 1 st , 0001 00:00:00, which has a serial datetime of -693593. The upper-limit is December 31 st , 9999, 23:59:59.999, which has a serial date-time of 2,958,465.9999884. The base date for this system is 00:00:00 on December 30 th , 1899, which has a serial date-time of 0.
> * In the 1904 date system, the lower limit is January 1 st , 0001, 00:00:00, which has a serial date-time of -695055. The upper limit is December 31 st , 9999, 23:59:59.999, which has a serial date-time of 2,957,003.9999884. The base date for this system is 00:00:00 on January 1 st , 1904, which has a serial date-time of 0.
>
> A serial date-time outside the temporal range for the selected date system is invalid.
> The date system is specified by the value of the date1904 attribute of the workbookPr element. [Example:
> 1900 date system: &lt;workbookPr showObjects="all"/&gt;
> 1904 date system: &lt;workbookPr date1904="1" showObjects="all"/&gt;
> end example]

Regarding the code structure - `Codec.Xlsx.Types.Common` looks like a suitable module for such helpers.
And thanks once again.

**Comment by qrilka** [comment]
Resolved by #95

**Comment by romildo** [comment]
@qrilka thank you.
